### ISSUE TITLE
restic backup fail

### ISSUE THREAD

**Issue by zaibon** [gigforks/minio#16]
I was playing a bit with restic and our minio see if it could be used as backend for restic.

first I created a restic repository on minio with
`restic  -r s3:http://195.134.212.42/restic/ init`
this worked fine, no problem

then I tried to backup a folder
` restic  -r s3:http://195.134.212.42/restic/ backup pictures/screenshots/`
Seems the backup went all the way but then crashed.

Restic logs:
```
scan [/home/zaibon/pictures/screenshots]
scanned 1 directories, 247 files in 0:00
[1:53] 100.00%  234.394 KiB/s  25.866 MiB / 25.866 MiB  248 / 248 items  0 errors  ETA 0:00 
duration: 1:53, 0.23MiB/s
unexpected EOF
```

Minio logs:
```
MakeBucketWithLocation bucket=restic, location=                                                                                                                                                [0/1138]
ListObjects bucket=restic, prefix=, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/keys/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/key/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/keys/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/key/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/keys/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/locks/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/locks/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/index/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/snapshots/, marker=, delimiter=, maxKeys=1000
StorageInfo
ListObjects bucket=restic, prefix=, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/keys/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/key/, marker=, delimiter=/, maxKeys=1000
ListObjects bucket=restic, prefix=restic/keys/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/locks/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/locks/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/index/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=restic/snapshots/, marker=, delimiter=, maxKeys=1000
ListObjects bucket=restic, prefix=, marker=, delimiter=/, maxKeys=1000
StorageInfo
         failed, startOffset != 0 =  1921058

Trace: 1: cmd/object-handlers.go:172:cmd.objectAPIHandlers.GetObjectHandler()
       2: cmd/api-router.go:59:cmd.(objectAPIHandlers).GetObjectHandler-fm()
       3: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
       4: vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP()
       5: cmd/generic-handlers.go:563:cmd.pathValidityHandler.ServeHTTP()
       6: cmd/generic-handlers.go:65:cmd.requestSizeLimitHandler.ServeHTTP()
       7: :0:()
       8: cmd/crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP()
       9: :0:()
      10: cmd/generic-handlers.go:219:cmd.redirectHandler.ServeHTTP()
      11: :0:()
      12: cmd/generic-handlers.go:277:cmd.minioReservedBucketHandler.ServeHTTP()
      13: :0:()
      14: cmd/generic-handlers.go:246:cmd.cacheControlHandler.ServeHTTP()
      15: :0:()
      16: cmd/generic-handlers.go:347:cmd.timeValidityHandler.ServeHTTP()
      17: :0:()
      18: vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1()
      19: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
      20: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
      21: cmd/generic-handlers.go:453:cmd.resourceHandler.ServeHTTP()
      22: cmd/auth-handler.go:245:cmd.authHandler.ServeHTTP()
      23: :0:()
      24: pkg/http/server.go:111:http.(*Server).Start.func1()
      25: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
      26: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
      27: net/http/server.go:2694:http.serverHandler.ServeHTTP()
      28: net/http/server.go:1830:http.(*conn).serve()
[2018-03-07T07:50:52.335261505Z] [ERROR] Unable to write to client. (Not Implemented)
2018/03/07 07:50:52 http: multiple response.WriteHeader calls
```

Not sure if that is a bug on our implementation of the gateway or just restic using some of the API calls that we dont support yet.

**Comment by iwanbk** [comment]
> failed, startOffset != 0 =  1921058

It is because of https://github.com/zero-os/0-stor/issues/499

**Comment by iwanbk** [comment]
> It is because of zero-os/0-stor#499

Fixed on https://github.com/zero-os/0-stor/pull/521, we could work on this issue now.

Start by reproducing it.

**Comment by iwanbk** [comment]
Fixed at b505ecd5cf0e3706bf09b5ea9e270eb37904a963 .

Tested using : 
- minio-go client to do read with range
- restic backup
```bash
$ restic -r s3:http://127.0.0.1:9000/bekap backup gateway
enter password for repository: 
password is correct
scan [/home/ibk/pathgo/src/github.com/minio/minio/cmd/gateway]
scanned 10 directories, 27 files in 0:00
[0:00] 100.00%  243.345 KiB / 243.345 KiB  37 / 37 items  0 errors  ETA 0:00 
duration: 0:00
snapshot 795fa854 saved
```

**Comment by zaibon** [comment]
I can't manage to get restic working on my end.
```
▶ restic  -r s3:http://195.134.212.42/restic/ backup ~/pictures/           
enter password for repository: 
scan [/home/zaibon/pictures]
scanned 2 directories, 257 files in 0:00
[0:43] 100.00%  625.661 KiB/s  26.273 MiB / 26.273 MiB  259 / 259 items  0 errors  ETA 0:00 
duration: 0:43, 0.60MiB/s
unexpected EOF
```
Stil the EOF error, but don't have any error in minio logs

**Comment by iwanbk** [comment]
$ restic version
restic 0.8.3
compiled with go1.10 on linux/amd64

**Comment by iwanbk** [comment]
It seems that 0-stor ReadRange used to fix this issue has some flaw, see https://github.com/zero-os/0-stor/issues/499#issuecomment-374141004.

**Comment by iwanbk** [comment]
I disabled the ReadRange until https://github.com/zero-os/0-stor/issues/499#issuecomment-374141004 fixed

**Comment by iwanbk** [comment]
> I disabled the ReadRange until zero-os/0-stor#499 (comment) fixed

I have the fix at https://github.com/zero-os/0-stor/pull/524, waiting for review/merge

**Comment by iwanbk** [comment]
fixed at f9204eea61877dc144bb4e6826831445b9a4099c
